### ISSUE TITLE
Problem: "Bad message"

### ISSUE THREAD

**Issue by ZRginger** [romanzaikin/BurpExtension-WhatsApp-Decryption-CheckPoint#47]
Hi! When I wanted to decrypt message, I got this information :" Bad message, can't decrypt the incoming  message." And I'm sure that the private key and public key and ref were correct. So I have no idea where is the issue.  

![image](https://user-images.githubusercontent.com/34329573/53331880-f0a1f180-392c-11e9-91c3-360cffdd1860.png)

Can some people help me?? @romanzaikin 

And I already try several times. It still doesn't work.  Here are details:
ref: {"ref":"1@fmw/dw6TcLIThMjjG/Evzd11yOcqRjU821ci8Ko+m1y1gHpAqDJ+k2xk","wid":"***********5@c.us","connected":true,"isResponse":"false","serverToken":"1@e5hswShfPfSKo6SeNCcVdJ44M1vlVMw3CkmBs+uadAGOb7oZAMAnlPMVl3x1n2cyn7dpSettMvJL/w==","browserToken":"1@zzZoKQj0hg/WwT/I/HYSxj/hgNXIR4eeP8fu+c+y/oItpzsrQKzeU0NKu01wRY8i48r8TiUBZz7kxqAJOh8zKn9KUyZt3Zm57jSRKa4YqBGWbA+M0TztsrGaFc5aEnRz6BrXUhAVrArcbjyxe6PAMw==","clientToken":"UW6FqwWt/779bcfVMauUEZ6V74KBsUx3eBJTIrygMHBIKncYNU5Slza0GjrwObWQwkQVhV0oKnqOmm27OWiwsA==","lc":"CN","lg":"zh","secret":"GPilAuRdetgPixBhJNN42Fouj3lDFhuL2QvlcGp800jf5J7RXPK00GYfJOpV5vOVaLOYGhsXL9YVqdGVgicV9ITv3Uy3I6iMauF0ixcwt3qV0AwhyFIIB6Bo7Kp0ZSk6iWCBbDC+aAbIwTJSczprwW56WCiQ61hvVC84ew2snazn2wQPW1ZxGFMTR61owmY6","protoVersion":[0,17],"binVersion":10,"battery":74,"plugged":false,"platform":"iphone","features":{"KEY_PARTICIPANT":true,"FLAGS":"EAEYASACKAIwAjgBQAFIAVgCYAF4AoABAogBAg=="},"phone":{"wa_version":"2.18.102","mcc":"460","mnc":"001","os_version":"12.1.2","device_manufacturer":"Apple","device_model":"iPhone 6s","os_build_number":"undefined"},"pushname":"Giant","tos":0}
priKey:
[140, 175, 120, 170, 163, 195, 128, 156, 29, 63, 241, 172, 97, 142, 249, 118, 50, 246, 186, 125, 179, 147, 102, 211, 53, 111, 229, 250, 144, 116, 13, 44]
pubKey:
[248, 172, 165, 193, 168, 46, 193, 66, 162, 197, 189, 212, 233, 92, 169, 11, 46, 57, 108, 228, 85, 248, 50, 198, 182, 196, 105, 204, 14, 169, 174, 87]

And converse incoming message in Base64.

**Comment by ZRginger** [comment]
Now I find that the message should be converted to base 64 encode. Just like this:
![image](https://user-images.githubusercontent.com/34329573/53314014-08f71980-38f7-11e9-8f73-7728e7b810e8.png)

**Comment by ZRginger** [comment]
But it still not work! Can anyone helps me?

**Comment by Orinion** [comment]
are you sure its Base64? It has to end with a "="

**Comment by ZRginger** [comment]
> are you sure its Base64? It has to end with a "="

Yes, I use Ctrl +B. Here is another example!
![image](https://user-images.githubusercontent.com/34329573/53344044-a9c3f400-394c-11e9-87cf-4f342aa1e0af.png)

**Comment by Orinion** [comment]
yes it works for me. Maybe you have to redo it because your keys aren't correct. I don't see annything else that is wrong

**Comment by ZRginger** [comment]
I want to cry now! Because I have done this at least 3 times, including  change keypair and change messages. Maybe I need to try again ... .. Anyway Thanks your help.

**Comment by ZRginger** [comment]
> yes it works for me. Maybe you have to redo it because your keys aren't correct. I don't see annything else that is wrong

Hi, I need your help! Can you help me to check if the info I filled into burp is wrong.
Here is the example info. And If you can please rely quickly, or I will delete the info.

ref:
{"ref":"1@t0csyK8UKDiHCn3dbPGaK6/QJqa18SJiLqil/kUzPxdzsLLWZlb7aO57","wid":"8613275918395@c.us","connected":true,"isResponse":"false","serverToken":"1@fTTx1ZQiZEHL6pLJbV7LkAvras4Qj7CA+ibvOTnXSdGrc+uegtR46KLVpC20kskZ5CwCUPhVQOv1VA==","browserToken":"1@/weUQ56/0TK4Rh5sFUZNTA8r7jZRzcEN6wIviLVuGEPJXnJhDfPYAY8+IfZAzASZxTfKlYfeO/nwf2YgFewBugxGaM0+ErxcQtQCvbFd6Kjr9w5BvnGzAVH/NhAdfZG64FYa9mICI+QRjdV/JhJApw==","clientToken":"1J7i4WUbAWt4pq6NAb38x9CaaENnhW5JcynDh0kW+lK5+VU6a0v9t+yRZDt2LJnimo8HJpS20VbX6IwcK+9sPA==","lc":"CN","lg":"zh","secret":"RPgj8QFD56Tg2L8OsQbo6DE7mc8d7YWubgHbu+RtZHpLCT8tEDKGF6fNTSGTQ65HymJ6uXTY8ZCFdkJ/2Ov0zUzOx6DiaksqYhv4Q+pkYvFkOzYtx2ePmfn6EyD9A4LD598uzq3fzxV2L/5iBxeQDLOhvEbEfX4hmJbGjwzHLtNo+wPuxeBoSU4SiUgYquj2","protoVersion":[0,17],"binVersion":10,"battery":44,"plugged":false,"platform":"iphone","features":{"KEY_PARTICIPANT":true,"FLAGS":"EAEYASACKAIwAjgBQAFIAVgCYAF4AoABAogBAg=="},"phone":{"wa_version":"2.18.102","mcc":"460","mnc":"001","os_version":"12.1.2","device_manufacturer":"Apple","device_model":"iPhone 6s","os_build_number":"undefined"},"pushname":"Giant","tos":0}

priKey:
[216, 154, 139, 152, 174, 239, 18, 78, 180, 68, 59, 170, 111, 168, 28, 148, 86, 41, 210, 127, 55, 144, 75, 67, 76, 210, 80, 106, 190, 149, 7, 110]

pubKey:
[251, 149, 138, 126, 173, 156, 150, 85, 17, 24, 120, 120, 249, 28, 58, 105, 25, 231, 184, 12, 131, 0, 249, 143, 171, 127, 79, 186, 27, 172, 250, 15]

imcoming message decryprt:
cHJlZW1wdC0xNTUxMTA3MDgwLTExLAWB0IXfLDAY4++VNwZ1Q/5Enig/0L4oFTVPNDOCyxp0ZW9+W0kbNH5b9umy1xhzFAqsJ/zUrDQBp6W+Shvas0fYvUAchRrLEEXdh4706q4Zs1873oW49rCtxdzV9ABkNxnARQTvWElMkGyxtQLJxiato3MN79Hx6EVB3U/I1gie6ZCYv/zFISk/+9YAT+lSHg==

**Comment by Orinion** [comment]
sorry, i don't know what is wrong

**Comment by ZRginger** [comment]
> sorry, i don't know what is wrong

Can you decrypt the message exactly? If you can, may be wrong in  my own project.

**Comment by Orinion** [comment]
i can't decrypt yours. however i can decrypt my own
![grafik](https://user-images.githubusercontent.com/12730894/53378524-2fa06900-3966-11e9-9df5-91fedf29ddd9.png)

**Comment by ZRginger** [comment]
> i can't decrypt yours. however i can decrypt my own
> ![grafik](https://user-images.githubusercontent.com/12730894/53378524-2fa06900-3966-11e9-9df5-91fedf29ddd9.png)

I copied the Unite8Array as keys:
![image](https://user-images.githubusercontent.com/34329573/53383759-ab74d280-39b3-11e9-9aed-2800bf299ff3.png)

And then I copied this as ref:
![image](https://user-images.githubusercontent.com/34329573/53383837-e2e37f00-39b3-11e9-9a4a-7656f82ee341.png)

 I converted this in base64, and decrypt the info:
![image](https://user-images.githubusercontent.com/34329573/53383922-25a55700-39b4-11e9-8756-614cf4256afb.png)

Finally, I got this:
![image](https://user-images.githubusercontent.com/34329573/53384030-7e74ef80-39b4-11e9-9a45-9a3bb9e5678f.png)

**Comment by ZRginger** [comment]
Hi! I find I  can decrypt some message, but other I can't. Can you tell me why? 
![image](https://user-images.githubusercontent.com/34329573/53384664-b4b36e80-39b6-11e9-93a8-8656521ae237.png)

![image](https://user-images.githubusercontent.com/34329573/53384747-efb5a200-39b6-11e9-8294-285db453e8c8.png)

**Comment by Orinion** [comment]
yes you cant decrypt every message yoou get. Thats because the others are not Chat messages.
The ones you can decrypt start with a hex number of length 16 which is followed by weird characters.
[this](https://user-images.githubusercontent.com/34329573/53383922-25a55700-39b4-11e9-8756-614cf4256afb.png) is not a message you can decrypt because it starts with preempt. I don't know much more, i hope i could help.

**Comment by ZRginger** [comment]
Thanks your help.

**Comment by baptx** [comment]
@Orinion @ZRginger for information, I can decrypt received preempt messages also with the script (https://pastebin.com/hNBeEqgv) I shared here: https://github.com/romanzaikin/BurpExtension-WhatsApp-Decryption-CheckPoint/issues/3
But it is not working to encrypt loaded or sent messages in a group for example. Are you able to modify and encrypt received / sent messages to a group?
I was not able to use the Burp Extension even when using Ctrl+B to get Base64 data, I got the "Bad message" error also but I know the secret and private / public key are correct because it worked with another script (https://pastebin.com/1Lb35yCD).
